platform/engineering/design-docs/2020-12-10-auto-deploy-cms-build.md

```diff
@@ -0,0 +1,216 @@
+
+As with the full build, during the rollout, there will be a comparison of GraphQL and CMS export builds to report errors when discrepancies between the two are found. The script used for this comparison will have some console output as well as an output JSON file describing the discrepancies.
+
+To determine whether a page exists in the Drupal content, we can go to [`staging.va.gov/drupal/debug/`](staging.va.gov/drupal/debug/). This page is not available in production, however.
+
+### Caveats
+
+To be determined.
+
+### Security Concerns
+
+There are no new security concerns with the auto-deploy. There is no user interaction involved unless we include a manual trigger for use by internal VSP engineers, which still should not present any threat of malicious requests. Denial of service is also not a concern as builds are queued.
+
+### Privacy Concerns
+
+There are no new privacy concerns with the auto-deploy as user data is not involved.
+
+### Open Questions and Risks
+
+#### Questions
+
+- What is the ideal schedule for the auto-deploy?
+  - The pipeline will initially run hourly 8am-8pm ET on weekdays.
+  - We want the highest rate of deploy that the system can sustain.
+
+- How should we coordinate the schedule of the auto-deploy with the daily production deploy?
+  - The deploys from both pipelines will likely overlap.
+  - The daily production deploy is still necessary for tagging the release.
+  - [The queue of content-only deploys should naturally resolve conflicts without any special handling.](https://github.com/department-of-veterans-affairs/va.gov-team/issues/18393)
+
+- Is there a way to get the time of the last updated published content in Drupal and use that to determine if we should build or not?
+  - If we have access to that information, we may not need to compare the CMS export tarballs.
+
+#### Risks
+
+- Newly added content from CMS without associated transformer schemas could fail the build.
+  - New content introduces the possibility of new discrepancies, which would fail any validations that depend on diffs between the GraphQL and CMS export builds.
+  - Having more content will inflate the runtime of the content-only deploy and could potentially exceed the scheduled interval.
+  - The best defense against this is accounting for null and empty fields in transformers.
+
+### Work Estimates
+
+- Create a new content-only deploy job for CMS export builds (5 pts)
+- Run the partial deploy on a schedule (3 pts)
+- Auto-deploy only new content (3 pts)
+
+### Alternatives
+
+Ideally, content would get deployed immediately as content writers make changes. However, the CMS export process is not suited for multiple invocations in quick succession, so we opted for a process that will essentially batch and deploy content changes as appropriate.
+
+### Future Work
+
+- There are plans to transition from Jenkins to CircleCI for our build pipeline. The auto-deploy job will have to be brought over to CircleCI at that time.
```

Comment:
Since this job requires access to the CMS drupal which is on the VA network, this job cannot be run from CircleCI, is my understanding. Operations _does_ plan to have a 'thing-doer' on the internal VA network. Another option would be to use Codebuild, which already exists as an AWS service, and can be triggered with a 'cron' job. The Ops group is currently determining what tool would be the 'thing-doer', there are a number of options but the requirement to be self-hosted is a limiter to those options. If anyone has any experience with a self-hosted 'thing-doer', I would be interested to hear about it.